### PR TITLE
Add optional ability to extend cluster lifetime.

### LINKED ISSUE #221: Add ability to extend cluster lifetime optionally
See https://bugzilla.mozilla.org/show_bug.cgi?id=1337809

### PR DESCRIPTION
**Pull request by jezdez**
Fix #221.

### CODE REVIEW

File: atmo/templates/atmo/keys/delete.html [hunk 1]
```diff
@@ -10,14 +10,14 @@
...
-    <form action="{% url 'keys-delete' id=key.id %}" method="POST" enctype="multipart/form-data">
+    <form action="{{ key.urls.delete }}" method="POST" enctype="multipart/form-data">
       {% csrf_token %}
       <button type="submit" class="btn btn-md btn-danger" title="Delete SSH key">
         <span class="glyphicon glyphicon-trash" aria-hidden="true"></span>
         <span class="submit-button">Delete key</span>
       </button>
       <button type="reset" class="btn btn-default btn-md hidden">Reset</button>
-      <a class="btn btn-default btn-md" href="{{ key.get_absolute_url }}">Cancel</a>
+      <a class="btn btn-default btn-md" href="{{ key.urls.detail }}">Cancel</a>
```
Comment (robhudson) [inline atmo/templates/atmo/keys/delete.html]: Just noticing this context uses `key` while others use `ssh_key`.

File: atmo/urlman.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    """
+    Special string subclass for URLs (for future with/without host modes)
+    """
+    def full(self, scheme='http', hostname='localhost', port='', params='',
+             query='', fragment='', secure=False):
+        netloc = hostname
+        if port:
+            netloc = '%s:%s' % (netloc, port)
+        if secure:
+            scheme = '%ss' % scheme
```
Comment (robhudson) [inline atmo/urlman.py]: Seems like this should check that `scheme` isn't already `"https"` when `secure=True`.

File: atmo/urlman.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    """Object which prints itself, essentially."""
+
+    def __init__(self, obj):
+        self.obj = obj
+
+    def __getattr__(self, attr):
+        return PrintMe(self.obj + '.' + attr)
+
+    def __str__(self):
+        return '{%s}' % self.obj
```
Comment (robhudson) [inline atmo/urlman.py]: Ok :)

File: tests/test_clusters.py [hunk 4]
```diff
@@ -194,9 +189,54 @@ def test_terminate_cluster(client, cluster_provisioner_mocks, cluster_factory,
...
+    assert cluster.lifetime_extension_count == 0
+    assert cluster.end_date is not None
+
+    cluster.extend(hours=3)
+    cluster.refresh_from_db()
+    assert cluster.lifetime_extension_count == 1
+    assert cluster.end_date > original_end_date
+    extended_end_date = cluster.end_date
+
+    # request that a new cluster be created
```
Comment (robhudson) [inline tests/test_clusters.py]: comment mismatch?

File: atmo/templates/atmo/keys/delete.html (hunk 1 above)
Comment (jezdez) [inline atmo/templates/atmo/keys/delete.html]: Good point, will fix.

File: atmo/urlman.py (hunk 2 above)
Comment (jezdez) [inline atmo/urlman.py]: good idea!

File: atmo/urlman.py (hunk 3 above)
Comment (jezdez) [inline atmo/urlman.py]: Hm, yeah, don't ask .)

File: tests/test_clusters.py (hunk 4 above)
Comment (jezdez) [inline tests/test_clusters.py]: Yep, copy/pasta

**Pull request merged**